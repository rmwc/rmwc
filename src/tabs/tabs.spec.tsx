import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { TabBar, Tab } from './';

describe('Tabs', () => {
  beforeAll(() => {
    // scrollTo is not implemented in JSDOM, mock it
    // @ts-ignore
    global.scrollTo = jest.fn();
  });

  it('TabBar renders', () => {
    const { asFragment } = render(
      <TabBar activeTabIndex={0} onActivate={(evt) => {}}>
        <Tab>Test</Tab>
      </TabBar>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Can be stacked', () => {
    const { asFragment } = render(
      <TabBar>
        <Tab stacked>Test</Tab>
      </TabBar>
    );
    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Can have label', () => {
    const { asFragment } = render(
      <TabBar>
        <Tab label="Test" />
      </TabBar>
    );
    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Can have icon', () => {
    const { asFragment } = render(
      <TabBar>
        <Tab icon="favorite" label="Test" />
      </TabBar>
    );
    expect(screen.getByText('favorite')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Can have restrictIndicator', () => {
    const { asFragment } = render(
      <TabBar>
        <Tab restrictIndicator icon="favorite" label="Test" />
      </TabBar>
    );
    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('can have no tabs', () => {
    const { asFragment } = render(<TabBar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('can have custom classnames', () => {
    [TabBar, Tab].forEach((Component: React.ComponentType<any>) => {
      const { container } = render(
        <Component className={'my-custom-classname'} />
      );
      expect(container.firstChild).toHaveClass('my-custom-classname');
    });
  });

  it('sets initial active tab', async () => {
    render(
      <TabBar activeTabIndex={1}>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabBar>
    );
    await waitFor(() =>
      expect(screen.getByText('2').parentElement?.parentElement).toHaveClass(
        'mdc-tab--active'
      )
    );
  });

  it('can be used as a controlled element', async () => {
    const { container } = render(
      <TabBar activeTabIndex={1}>
        <Tab>1</Tab>
        <Tab>2</Tab>
      </TabBar>
    );

    userEvent.click(screen.getByText('1'));

    await waitFor(() =>
      expect(container.getElementsByClassName('mdc-tab--active')).toHaveLength(
        1
      )
    );
  });

  it('focuses active tab on mount', async () => {
    const { container } = render(
      <TabBar>
        <Tab focusOnActivate>Test</Tab>
      </TabBar>
    );

    return new Promise((resolve) => {
      window.requestAnimationFrame(() => {
        expect(document.activeElement).toBe(container.querySelector('button'));
        resolve(true);
      });
    });
  });

  it('does not focus active tab on mount', () => {
    const { container } = render(
      <TabBar>
        <Tab focusOnActivate={false}>Test</Tab>
      </TabBar>
    );

    expect(document.activeElement).not.toBe(container.querySelector('button'));
  });
});
