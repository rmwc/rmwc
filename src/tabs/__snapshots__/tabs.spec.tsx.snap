// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Can be stacked 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        >
          <button
            class="mdc-tab mdc-tab--stacked"
          >
            <div
              class="mdc-tab__content"
            >
              <span
                class="mdc-tab__text-label"
              >
                Test
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Tabs Can have icon 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        >
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <i
                class="rmwc-icon rmwc-icon--ligature material-icons mdc-tab__icon"
              >
                favorite
              </i>
              <span
                class="mdc-tab__text-label"
              >
                Test
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Tabs Can have label 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        >
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <span
                class="mdc-tab__text-label"
              >
                Test
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Tabs Can have restrictIndicator 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        >
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <i
                class="rmwc-icon rmwc-icon--ligature material-icons mdc-tab__icon"
              >
                favorite
              </i>
              <span
                class="mdc-tab__text-label"
              >
                Test
              </span>
              <span
                class="mdc-tab-indicator"
              >
                <span
                  aria-hidden="true"
                  class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
                />
              </span>
            </div>
            <div
              class="mdc-tab__ripple "
            />
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Tabs TabBar renders 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        >
          <button
            class="mdc-tab"
          >
            <div
              class="mdc-tab__content"
            >
              <span
                class="mdc-tab__text-label"
              >
                Test
              </span>
            </div>
            <span
              class="mdc-tab-indicator"
            >
              <span
                aria-hidden="true"
                class="mdc-tab-indicator__content mdc-tab-indicator__content--underline"
              />
            </span>
            <div
              class="mdc-tab__ripple "
            />
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Tabs can have no tabs 1`] = `
<DocumentFragment>
  <nav
    class="mdc-tab-bar"
  >
    <div
      class="mdc-tab-scroller"
    >
      <div
        class="mdc-tab-scroller__scroll-area mdc-tab-scroller__scroll-area--scroll"
        style="margin-bottom: 0px;"
      >
        <div
          class="mdc-tab-scroller__scroll-content"
        />
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
